File: src/test/scala/com/exasol/cloudetl/parquet/ParquetRowReaderPrimitiveTypesTest.scala
package com.exasol.cloudetl.parquet

import java.math.BigDecimal
import java.math.BigInteger
import java.math.MathContext
import java.sql.Timestamp

import com.exasol.cloudetl.parquet.converter.ParquetDecimalConverter
import com.exasol.common.data.Row

import org.apache.parquet.column.Dictionary
import org.apache.parquet.example.data.simple.SimpleGroup
import org.apache.parquet.io.api.Binary
import org.apache.parquet.schema.LogicalTypeAnnotation.decimalType
import org.apache.parquet.schema.MessageTypeParser
import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName
import org.apache.parquet.schema.Type.Repetition
import org.apache.parquet.schema.Types

class ParquetRowReaderPrimitiveTypesTest extends BaseParquetReaderTest {

  test("reads INT64 (TIMESTAMP_MILLIS) as timestamp value") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int64 col_long;
         |  required int64 col_timestamp (TIMESTAMP_MILLIS);
         |}
         |""".stripMargin
    )
    val timestamp = new Timestamp(System.currentTimeMillis())
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_long", 153L)
      record.append("col_timestamp", timestamp.getTime())
      writer.write(record)
    }
    assert(getRecords()(0) === Row(Seq(153L, timestamp)))
  }

  test("reads FIXED_LEN_BYTE_ARRAY as string value") {
    val size = 5
    val schema = MessageTypeParser.parseMessageType(
      s"""|message test {
          |  required fixed_len_byte_array($size) col_byte_array;
          |}
          |""".stripMargin
    )
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_byte_array", "hello")
      writer.write(record)
    }
    assert(getRecords()(0) === Row(Seq("hello")))
  }

  test("reads BINARY as string value") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required binary col_binary;
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_binary", "test")
      writer.write(record)
    }
    assert(getRecords()(0) === Row(Seq("test")))
  }

  test("reads BINARY (UTF8) as string value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required binary col_binary (UTF8);
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq("test1", "test2", "test1", "test2").foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("col_binary", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === Row(Seq("test1")))
  }

  test("reads INT32 (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int32 col_int (DECIMAL(9,2));
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(123456789, 0, 0, 123456789).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("col_int", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === Row(Seq(BigDecimal.valueOf(123456789, 2))))
  }

  test("reads INT64 (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int64 column (DECIMAL(18,2));
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(1234567890123456L, 1L, 1L, 1234567890123456L).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("column", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === Row(Seq(BigDecimal.valueOf(1234567890123456L, 2))))
  }

  test("reads BINARY (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required binary column (DECIMAL(30,2));
         |}
         |""".stripMargin
    )
    val decimalValue = "123456789012345678901234567890"
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(decimalValue, decimalValue).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("column", value)
        writer.write(record)
      }
    }
    val expected = new BigDecimal(new BigInteger(decimalValue.getBytes()), 2, new MathContext(30))
    val records = getRecords()
    assert(records.size === 2)
    assert(records(0) === Row(Seq(expected)))
  }

  test("reads FIXED_LEN_BYTE_ARRAY (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required fixed_len_byte_array(20) column (DECIMAL(20,2));
         |}
         |""".stripMargin
    )
    val decimalValue = "12345678901234567890"
    val bytes = Array.fill[Byte](30)(0x1)
    val zeros = Binary.fromConstantByteArray(bytes, 0, 20)
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(decimalValue, decimalValue).foreach { value =>
        var record = new SimpleGroup(schema)
        record.append("column", value)
        writer.write(record)

        record = new SimpleGroup(schema)
        record.append("column", zeros)
        writer.write(record)
      }
    }
    val expected = new BigDecimal(new BigInteger(decimalValue.getBytes()), 2, new MathContext(20))
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === Row(Seq(expected)))
  }

  test("read throws when decimal converter has unsupported base primitive type") {
    final case class DictionaryEncoding() extends Dictionary(null) {
      override def getMaxId(): Int = 4
    }
    val parquetType = Types
      .primitive(PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY, Repetition.OPTIONAL)
      .length(4)
      .as(decimalType(0, 9))
      .named("bytes")
    val thrown = intercept[UnsupportedOperationException] {
      ParquetDecimalConverter(parquetType, 0, null).setDictionary(DictionaryEncoding())
    }
    assert(thrown.getMessage.contains("Cannot convert parquet type to decimal type"))
  }

}

File: src/test/java/com/exasol/cloudetl/parquet/ParquetRowReaderPrimitiveTypesTest.java
package com.exasol.cloudetl.parquet;

import static org.junit.jupiter.api.Assertions.*;

import java.math.*;
import java.nio.charset.StandardCharsets;
import java.sql.Timestamp;
import java.util.*;

import org.apache.parquet.example.data.simple.SimpleGroup;
import org.apache.parquet.io.api.Binary;
import org.apache.parquet.schema.MessageTypeParser;
import org.junit.jupiter.api.Test;

import com.exasol.cloudetl.helper.UUIDConverter;
import com.exasol.common.data.Row;

class ParquetRowReaderPrimitiveTypesTest extends BaseParquetReaderTest {
    @Test
    void readsInt64TimestampMillisAsTimestampValue() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required int64 col_long;\n" +
"  required int64 col_timestamp (TIMESTAMP_MILLIS);\n" +
"}\n" +
"\n"));
        final Timestamp timestamp = new Timestamp(System.currentTimeMillis());
        withResource(getParquetWriter(schema, false), writer -> {
            final SimpleGroup record = new SimpleGroup(schema);
            record.append("col_long", 153L);
            record.append("col_timestamp", timestamp.getTime());
            writer.write(record);
        });
        assertEquals(row(153L, timestamp), getRecords().get(0));
    }

    @Test
    void readsInt64TimestampMicrosAsTimestampValue() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required int64 col_timestamp (TIMESTAMP_MICROS);\n" +
"}\n" +
"\n"));
        final Timestamp timestamp = Timestamp.valueOf("2022-01-12 08:28:53.123456");
        final long micros = timestamp.getTime() * 1000L + timestamp.getNanos() / 1000L % 1000L;
        withResource(getParquetWriter(schema, false), writer -> {
            final SimpleGroup record = new SimpleGroup(schema);
            record.append("col_timestamp", micros);
            writer.write(record);
        });
        assertEquals(row(timestamp), getRecords().get(0));
    }

    @Test
    void readsFixedLenByteArrayAsStringValue() throws Exception {
        final int size = 5;
        final var schema = MessageTypeParser.parseMessageType(String.format("message test {\n" +
"  required fixed_len_byte_array(%d) col_byte_array;\n" +
"}\n" +
"\n", size));
        withResource(getParquetWriter(schema, false), writer -> {
            final SimpleGroup record = new SimpleGroup(schema);
            record.append("col_byte_array", "hello");
            writer.write(record);
        });
        assertEquals(row("hello"), getRecords().get(0));
    }

    @Test
    void readsBinaryAsStringValue() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required binary col_binary;\n" +
"}\n" +
"\n"));
        withResource(getParquetWriter(schema, false), writer -> {
            final SimpleGroup record = new SimpleGroup(schema);
            record.append("col_binary", "test");
            writer.write(record);
        });
        assertEquals(row("test"), getRecords().get(0));
    }

    @Test
    void readsBinaryUtf8AsStringValueUsingDictionaryEncoding() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required binary col_binary (UTF8);\n" +
"}\n" +
"\n"));
        withResource(getParquetWriter(schema, true), writer -> {
            for (final String value : List.of("test1", "test2", "test1", "test2")) {
                final SimpleGroup record = new SimpleGroup(schema);
                record.append("col_binary", value);
                writer.write(record);
            }
        });
        final List<Row> records = getRecords();
        assertEquals(4, records.size());
        assertEquals(row("test1"), records.get(0));
    }

    @Test
    void readsInt32DecimalAsBigDecimalValueUsingDictionaryEncoding() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required int32 col_int (DECIMAL(9,2));\n" +
"}\n" +
"\n"));
        withResource(getParquetWriter(schema, true), writer -> {
            for (final int value : List.of(123456789, 0, 0, 123456789)) {
                final SimpleGroup record = new SimpleGroup(schema);
                record.append("col_int", value);
                writer.write(record);
            }
        });
        final List<Row> records = getRecords();
        assertEquals(4, records.size());
        assertEquals(row(BigDecimal.valueOf(123456789, 2)), records.get(0));
        assertEquals(row(BigDecimal.valueOf(0, 2)), records.get(1));
    }

    @Test
    void readsInt64DecimalAsBigDecimalValueUsingDictionaryEncoding() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required int64 column (DECIMAL(18,2));\n" +
"}\n" +
"\n"));
        withResource(getParquetWriter(schema, true), writer -> {
            for (final long value : List.of(1234567890123456L, 1L, 1L, 1234567890123456L)) {
                final SimpleGroup record = new SimpleGroup(schema);
                record.append("column", value);
                writer.write(record);
            }
        });
        final List<Row> records = getRecords();
        assertEquals(4, records.size());
        assertEquals(row(BigDecimal.valueOf(1234567890123456L, 2)), records.get(0));
        assertEquals(row(BigDecimal.valueOf(1L, 2)), records.get(1));
    }

    @Test
    void readsBinaryDecimalAsBigDecimalValueUsingDictionaryEncoding() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required binary column (DECIMAL(30,2));\n" +
"}\n" +
"\n"));
        final String decimalValue = "123456789012345678901234567890";
        withResource(getParquetWriter(schema, true), writer -> {
            for (final String value : List.of(decimalValue, decimalValue)) {
                final SimpleGroup record = new SimpleGroup(schema);
                record.append("column", value);
                writer.write(record);
            }
        });
        final BigDecimal expected = new BigDecimal(new BigInteger(decimalValue.getBytes(StandardCharsets.UTF_8)), 2,
                new MathContext(30));
        final List<Row> records = getRecords();
        assertEquals(2, records.size());
        assertEquals(row(expected), records.get(0));
    }

    @Test
    void readsFixedLenByteArrayDecimalAsBigDecimalValueUsingDictionaryEncoding() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required fixed_len_byte_array(9) column (DECIMAL(20,2));\n" +
"}\n" +
"\n"));
        final String decimalValue = "12345678901234567890";
        final Binary decimalBinary = Binary.fromConstantByteArray(new BigDecimal(decimalValue).unscaledValue().toByteArray());
        final Binary zeros = Binary.fromConstantByteArray(new byte[9], 0, 9);
        withResource(getParquetWriter(schema, true), writer -> {
            for (final Binary value : List.of(decimalBinary, zeros)) {
                final SimpleGroup record = new SimpleGroup(schema);
                record.append("column", value);
                writer.write(record);
            }
        });
        final BigDecimal expected = new BigDecimal(new BigInteger(decimalValue), 2, new MathContext(20));
        final List<Row> records = getRecords();
        assertEquals(2, records.size());
        assertEquals(row(expected), records.get(0));
        assertEquals(row(BigDecimal.valueOf(0, 2)), records.get(1));
    }

    @Test
    void readsFixedLenByteArrayUuidAsStringValue() throws Exception {
        final var schema = MessageTypeParser.parseMessageType(("message test {\n" +
"  required fixed_len_byte_array(16) column (UUID);\n" +
"}\n" +
"\n"));
        final UUID uuid = UUID.randomUUID();
        withResource(getParquetWriter(schema, true), writer -> {
            final SimpleGroup record = new SimpleGroup(schema);
            record.append("column", Binary.fromConstantByteArray(UUIDConverter.toByteArray(uuid)));
            writer.write(record);
        });
        assertEquals(row(uuid.toString()), getRecords().get(0));
    }
}
